From 7eae12319512493a7b4b33f6a33f4fbef40e7caf Mon Sep 17 00:00:00 2001
From: Victor Bo <bvoid@yandex.ru>
Date: Sun, 16 Feb 2020 22:25:00 +0200
Subject: [PATCH 2/7] Add double tap to sleep gesture

  * based on
  https://github.com/crdroidandroid/android_packages_apps_Launcher3/commit/1ced8c431b3a8ccb397199f68e1dd1ac401a90d4
  https://github.com/crdroidandroid/android_packages_apps_Launcher3/commit/2581d966c61bbfca344803eb443f5426abf2348a

	modified:   Android.mk
	modified:   AndroidManifest-common.xml
	modified:   AndroidManifest.xml
	modified:   SecondaryDisplayLauncher/Android.mk
	modified:   res/values/strings.xml
	modified:   res/xml/launcher_preferences.xml
	modified:   src/com/android/launcher3/Utilities.java
	modified:   src/com/android/launcher3/Workspace.java
	modified:   src/com/android/launcher3/touch/WorkspaceTouchListener.java

Change-Id: I9a172cfbc38617914dcdc626e19e397ca8d7c229
---
 Android.mk                                    | 10 +++++++---
 AndroidManifest-common.xml                    |  6 +++++-
 AndroidManifest.xml                           |  4 +++-
 res/values-ru/strings.xml                     |  2 ++
 res/values/strings.xml                        |  4 ++++
 res/xml/launcher_preferences.xml              |  7 +++++++
 src/com/android/launcher3/Utilities.java      |  7 +++++++
 src/com/android/launcher3/Workspace.java      | 20 +++++++++++++++++++
 .../touch/WorkspaceTouchListener.java         |  1 +
 9 files changed, 56 insertions(+), 5 deletions(-)

diff --git a/Android.mk b/Android.mk
index e3daddc2a..50c730691 100644
--- a/Android.mk
+++ b/Android.mk
@@ -60,7 +60,8 @@ LOCAL_PROTOC_OPTIMIZE_TYPE := nano
 LOCAL_PROTOC_FLAGS := --proto_path=$(LOCAL_PATH)/protos/ --proto_path=$(LOCAL_PATH)/proto_overrides/
 LOCAL_PROTO_JAVA_OUTPUT_PARAMS := enum_style=java
 
-LOCAL_SDK_VERSION := current
+LOCAL_PRIVATE_PLATFORM_APIS := true
+LOCAL_CERTIFICATE := platform
 LOCAL_MIN_SDK_VERSION := 26
 LOCAL_MODULE := Launcher3CommonDepsLib
 LOCAL_PRIVILEGED_MODULE := true
@@ -89,7 +90,8 @@ LOCAL_PROGUARD_FLAG_FILES := proguard.flags
 # Proguard is disable for testing. Derivarive prjects to keep proguard enabled
 LOCAL_PROGUARD_ENABLED := disabled
 
-LOCAL_SDK_VERSION := current
+LOCAL_PRIVATE_PLATFORM_APIS := true
+LOCAL_CERTIFICATE := platform
 LOCAL_MIN_SDK_VERSION := 26
 LOCAL_PACKAGE_NAME := Trebuchet
 LOCAL_PRIVILEGED_MODULE := true
@@ -120,7 +122,8 @@ LOCAL_RESOURCE_DIR := $(LOCAL_PATH)/go/res
 
 LOCAL_PROGUARD_FLAG_FILES := proguard.flags
 
-LOCAL_SDK_VERSION := current
+LOCAL_PRIVATE_PLATFORM_APIS := true
+LOCAL_CERTIFICATE := platform
 LOCAL_MIN_SDK_VERSION := 26
 LOCAL_PACKAGE_NAME := TrebuchetGo
 LOCAL_PRIVILEGED_MODULE := true
@@ -191,6 +194,7 @@ else
   LOCAL_MIN_SDK_VERSION := 26
 endif
 LOCAL_PACKAGE_NAME := TrebuchetQuickStep
+LOCAL_CERTIFICATE := platform
 LOCAL_PRIVILEGED_MODULE := true
 LOCAL_SYSTEM_EXT_MODULE := true
 LOCAL_OVERRIDES_PACKAGES := Home Launcher2 Launcher3 Launcher3QuickStep
diff --git a/AndroidManifest-common.xml b/AndroidManifest-common.xml
index 2ba068f58..2b0dc8e6f 100644
--- a/AndroidManifest-common.xml
+++ b/AndroidManifest-common.xml
@@ -19,7 +19,9 @@
 -->
 <manifest
     xmlns:android="http://schemas.android.com/apk/res/android"
-    package="com.android.launcher3">
+    package="com.android.launcher3"
+    coreApp="true"
+    android:sharedUserId="android.uid.system">
 
     <!--
     The manifest defines the common entries that should be present in any derivative of Launcher3.
@@ -47,6 +49,8 @@
     <uses-permission android:name="android.permission.QUERY_ALL_PACKAGES" />
     <uses-permission android:name="android.permission.USE_BIOMETRIC" />
 
+    <uses-permission android:name="android.permission.DEVICE_POWER" />
+
     <!-- TODO(b/150802536): Enabled only for ENABLE_FIXED_ROTATION_TRANSFORM feature flag -->
     <uses-permission android:name="android.permission.WRITE_SECURE_SETTINGS"/>
     
diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index fcad766ae..bd9eb5a9f 100644
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -20,7 +20,9 @@
 <manifest
     xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools"
-    package="com.android.launcher3">
+    package="com.android.launcher3"
+    coreApp="true"
+    android:sharedUserId="android.uid.system">
     <uses-sdk android:targetSdkVersion="29" android:minSdkVersion="26"/>
     <!--
     Manifest entries specific to Launcher3. This is merged with AndroidManifest-common.xml.
diff --git a/res/values-ru/strings.xml b/res/values-ru/strings.xml
index 4d8ba8860..0d90a0d8f 100644
--- a/res/values-ru/strings.xml
+++ b/res/values-ru/strings.xml
@@ -147,4 +147,6 @@
     <string name="work_apps_paused_content_description" msgid="7553586952985486433">"Рабочий профиль приостановлен. Рабочие приложения не могут отправлять уведомления, расходовать заряд батареи и получать доступ к данным о вашем местоположении."</string>
     <string name="work_switch_tip" msgid="808075064383839144">"Приостановить рабочие приложения и уведомления"</string>
     <string name="remote_action_failed" msgid="1383965239183576790">"Не удалось выполнить действие (<xliff:g id="WHAT">%1$s</xliff:g>)."</string>
+    <string name="pref_sleep_gesture_title">"Двойное касание для перехода в сон"</string>
+    <string name="pref_sleep_gesture_summary">"Дважды быстро коснитесь пустого пространства для выключения экрана"</string>
 </resources>
diff --git a/res/values/strings.xml b/res/values/strings.xml
index 935bb40fb..1c2693ee3 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -355,4 +355,8 @@
 
     <!-- Failed action error message: e.g. Failed: Pause -->
     <string name="remote_action_failed">Failed: <xliff:g id="what" example="Pause">%1$s</xliff:g></string>
+
+    <!-- DT2S option -->
+    <string name="pref_sleep_gesture_title">Double tap to sleep</string>
+    <string name="pref_sleep_gesture_summary">Double tap on empty space for power-off screen</string>
 </resources>
diff --git a/res/xml/launcher_preferences.xml b/res/xml/launcher_preferences.xml
index 36eb21363..c95f6def4 100644
--- a/res/xml/launcher_preferences.xml
+++ b/res/xml/launcher_preferences.xml
@@ -45,6 +45,13 @@
         android:defaultValue="true"
         android:persistent="true" />
 
+    <SwitchPreference
+        android:key="pref_sleep_gesture"
+        android:title="@string/pref_sleep_gesture_title"
+        android:summary="@string/pref_sleep_gesture_summary"
+        android:defaultValue="true"
+        android:persistent="true"/>
+
     <SwitchPreference
         android:key="pref_allowRotation"
         android:title="@string/allow_rotation_title"
diff --git a/src/com/android/launcher3/Utilities.java b/src/com/android/launcher3/Utilities.java
index b32b39618..4201b535a 100644
--- a/src/com/android/launcher3/Utilities.java
+++ b/src/com/android/launcher3/Utilities.java
@@ -140,6 +140,9 @@ public final class Utilities {
     public static final String EXTRA_WALLPAPER_OFFSET = "com.android.launcher3.WALLPAPER_OFFSET";
     public static final String EXTRA_WALLPAPER_FLAVOR = "com.android.launcher3.WALLPAPER_FLAVOR";
 
+    // DT2S
+    public static final String SLEEP_GESTURE = "pref_sleep_gesture";
+
     public static boolean IS_RUNNING_IN_TEST_HARNESS =
                     ActivityManager.isRunningInTestHarness();
 
@@ -695,4 +698,8 @@ public final class Utilities {
         SharedPreferences prefs = getPrefs(context.getApplicationContext());
         return prefs.getBoolean(InvariantDeviceProfile.KEY_WORKSPACE_EDIT, true);
     }
+
+    public static boolean useSleepGesture(Context context) {
+        return getPrefs(context).getBoolean(SLEEP_GESTURE, true);
+    }
 }
diff --git a/src/com/android/launcher3/Workspace.java b/src/com/android/launcher3/Workspace.java
index 33ecd6b93..ad513a253 100644
--- a/src/com/android/launcher3/Workspace.java
+++ b/src/com/android/launcher3/Workspace.java
@@ -50,6 +50,7 @@ import android.graphics.drawable.Drawable;
 import android.os.Handler;
 import android.os.Message;
 import android.os.Parcelable;
+import android.os.PowerManager;
 import android.os.UserHandle;
 import android.text.TextUtils;
 import android.util.AttributeSet;
@@ -57,6 +58,7 @@ import android.util.Log;
 import android.util.SparseArray;
 import android.view.LayoutInflater;
 import android.view.MotionEvent;
+import android.view.GestureDetector;
 import android.view.View;
 import android.view.ViewGroup;
 import android.view.ViewTreeObserver;
@@ -252,6 +254,7 @@ public class Workspace extends PagedView<WorkspacePageIndicator>
 
     // Handles workspace state transitions
     private final WorkspaceStateTransitionAnimation mStateTransitionAnimation;
+    private GestureDetector mGestureListener;
 
     private final StatsLogManager mStatsLogManager;
 
@@ -286,10 +289,27 @@ public class Workspace extends PagedView<WorkspacePageIndicator>
 
         // Disable multitouch across the workspace/all apps/customize tray
         setMotionEventSplittingEnabled(true);
+
+        final PowerManager pm = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
+
+        mGestureListener =
+                new GestureDetector(context, new GestureDetector.SimpleOnGestureListener() {
+            @Override
+            public boolean onDoubleTap(MotionEvent event) {
+                if (Utilities.useSleepGesture(context)) {
+                    pm.goToSleep(event.getEventTime());
+                }
+                return true;
+            }
+        });
         setOnTouchListener(new WorkspaceTouchListener(mLauncher, this));
         mStatsLogManager = StatsLogManager.newInstance(context);
     }
 
+    public boolean checkDoubleTap(MotionEvent ev) {
+        return mGestureListener.onTouchEvent(ev);
+    }
+
     @Override
     public void setInsets(Rect insets) {
         DeviceProfile grid = mLauncher.getDeviceProfile();
diff --git a/src/com/android/launcher3/touch/WorkspaceTouchListener.java b/src/com/android/launcher3/touch/WorkspaceTouchListener.java
index 4fa658e7d..500a94c77 100644
--- a/src/com/android/launcher3/touch/WorkspaceTouchListener.java
+++ b/src/com/android/launcher3/touch/WorkspaceTouchListener.java
@@ -82,6 +82,7 @@ public class WorkspaceTouchListener extends GestureDetector.SimpleOnGestureListe
     @Override
     public boolean onTouch(View view, MotionEvent ev) {
         mGestureDetector.onTouchEvent(ev);
+        mWorkspace.checkDoubleTap(ev);
 
         int action = ev.getActionMasked();
         if (action == ACTION_DOWN) {
-- 
2.30.2

